src/TokenHandler/AccountManager.mo

```diff
@@ -0,0 +1,408 @@
+
+    /// Manages deposit balances for each user.
+    let depositRegistry : DepositRegistry.DepositRegistry = DepositRegistry.DepositRegistry(freezeCallback);
+
+    /// Manages acklog of principals waiting for processing.
+    let backlog : Backlog.Backlog = Backlog.Backlog();
+
+    /// Current fee amount.
+    var fee_ : Nat = 0;
+
+    /// Total deposited funds across all accounts.
+    var depositedFunds_ : Nat = 0;
+
+    /// Total amount consolidated. Accumulated value.
+    var totalConsolidated_ : Nat = 0;
+
+    /// Total amount withdrawn. Accumulated value.
+    var totalWithdrawn_ : Nat = 0;
+
+    /// Total funds queued for consolidation.
+    var queuedFunds : Nat = 0;
+
+    /// Total funds underway for consolidation.
+    var underwayFunds : Nat = 0;
+
+    /// Retrieves the current fee amount.
+    public func fee() : Nat = fee_;
+
+    /// Updates the fee amount based on the ICRC1 ledger.
+    public func updateFee() : async* Nat {
+      let newFee = await icrc1Ledger.icrc1_fee();
+      if (fee_ != newFee) {
+        journal.push((Time.now(), ownPrincipal, #feeUpdated({ old = fee_; new = newFee })));
+        fee_ := newFee;
+        recalculateBacklog();
+      };
+      newFee;
+    };
+
+    /// Retrieves the sum of all current deposits. This value is nearly the same as backlogFunds(), but includes
+    /// entries, which could not be added to backlog, for instance when balance less than fee.
+    /// It's always >= backlogFunds()
+    public func depositedFunds() : Nat = depositedFunds_;
+
+    /// Retrieves the sum of all successful consolidations.
+    public func totalConsolidated() : Nat = totalConsolidated_;
+
+    /// Retrieves the sum of all deductions from the main account.
+    public func totalWithdrawn() : Nat = totalWithdrawn_;
+
+    /// Retrieves the calculated balance of the main account.
+    public func consolidatedFunds() : Nat = totalConsolidated_ - totalWithdrawn_;
+
+    /// Returns the size of the consolidation backlog.
+    public func backlogSize() : Nat = backlog.size();
+
+    /// Returns the sum of all deposits in the backlog.
+    public func backlogFunds() : Nat = queuedFunds + underwayFunds;
+
+    /// Notifies of a deposit and schedules backlog processing.
+    /// Returns the newly detected deposit if successful.
+    public func notify(p : Principal) : async* ?Nat {
+      if (isFrozen() or not depositRegistry.lock(p)) return null;
+
+      let latestDeposit = try {
+        await* loadDeposit(p);
+      } catch (err) {
+        depositRegistry.unlock(p);
+        throw err;
+      };
+
+      depositRegistry.unlock(p);
+
+      let prevDeposit = updateDeposit(p, latestDeposit);
+
+      if (latestDeposit < prevDeposit) freezeCallback("latestDeposit < prevDeposit on notify");
+
+      if (latestDeposit > fee_) {
+        // precredit deposit funds
+        ignore creditRegistry.credit(p, latestDeposit - fee_);
```

Comment:
This seems wrong. In general, we should credit only `latestDeposit - prevDeposit` because `prevDeposit` was already credited before. The function `creditRegistry.credit` increments, not overwrites, so we have to credit the incremental difference. 

If `prevDeposit >= fee_` then we credit `latestDeposit - prevDeposit`. If `prevDeposit < fee_` then we credit `latestDeposit - fee_`.

Reply:
Sure, I will fix it